File: 017-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
# -*- coding:utf-8 -*-


# Given a digit string, return all possible letter combinations that the number could represent.
#
#
#
# A mapping of digit to letters (just like on the telephone buttons) is given below.
#
#
#
# Input:Digit string "23"
# Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
#
#
#
# Note:
# Although the above answer is in lexicographical order, your answer could be in any order you want.


class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if digits == '':
            return []
        
        dic = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
               
        return reduce(lambda accu, digit: [x + y for x in accu for y in dic[digit]], digits, [''])
        

File: 017-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.js
// Given a digit string, return all possible letter combinations that the number could represent.
//
//
//
// A mapping of digit to letters (just like on the telephone buttons) is given below.
//
//
//
// Input:Digit string "23"
// Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
//
//
//
// Note:
// Although the above answer is in lexicographical order, your answer could be in any order you want.


/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    const dict = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
    
    if (!digits.length) return []
    
    let res = dict[digits[0]].split('')
    digits.split('').slice(1).forEach(d => {
        const len = res.length
        const originalRes = [...res]
        
        // expand the original result
        for (let j = 0; j < dict[parseInt(d)].length - 1; j++) {
            res = res.concat(originalRes)
        }
    
        // append the character
        res = res.map((s, k) => {
            return s + dict[parseInt(d)][parseInt(k / len)]
        })
    })

    return res
};

